netlogo-gui/src/main/app/ChangeManager.scala

```diff
@@ -0,0 +1,138 @@
+package org.nlogo.app
+
+import java.util.List
+import java.util.ArrayList
+import javax.swing.event.ChangeEvent
+/**
+  * Manages a Queue of Changables to perform undo and/or redo operations. Clients can add implementations of the Changeable
+  * class to this class, and it will manage undo/redo as a Queue.
+  *
+  * @author Greg Cope
```

Comment:
http://www.algosome.com/articles/implementing-undo-redo-java.html
I went through their copyright page just now, looks pretty tight. May be you can go through that and understand it better than I did.